### PR TITLE
Zoom implementation of AspNet.Security.OAuth

### PR DESCRIPTION
**Pull request by softwareolatomiwa**
Zoom implementation of AspNet.Security.OAuth

### CODE REVIEW

**Comment by softwareolatomiwa** [comment]
Hi @martincostello
Not sure why this failed as it builds perfectly in my environment. Do you mind pointing out anything I need to do to make this pass

File: docs/zoom.md
Comment (martincostello) [inline docs/zoom.md:1]: If the provider doesn't do anything special, then we don't really need a documentation file. They exist to document additional settings beyond the base OAuth implementation.

File: src/AspNet.Security.OAuth.Zoom/AspNet.Security.OAuth.Zoom.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
+  </PropertyGroup>
+
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zoom/AspNet.Security.OAuth.Zoom.csproj]: This is needed to fix the build:

```suggestion
  <PropertyGroup>
    <PackageValidationBaselineVersion>8.0.1</PackageValidationBaselineVersion>
    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
  </PropertyGroup>

  <!-- TODO Enable once this provider is published to NuGet.org -->
  <PropertyGroup>
    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
  </PropertyGroup>

```

File: src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationConstants.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    /// <summary>
+    /// Available profile fields after a Zoom authentication.
+    /// See <a>https://developers.zoom.us/docs/api/rest/reference/user/methods/#operation/user</a>
+    /// </summary>
+    public static class ProfileFields
+    {
+        /// <summary>
+        /// The unique identifier for the given memembermber. May also be referenced as the <c>personId</c> within a Person URN (<c>urn:li:person:{personId}</c>).
+        /// The <c>id</c> is unique to your specific developer application. Any attempts to use the <c>id</c> with other developer applications will not succeed.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationConstants.cs]: This looks like copy-paste from the LinkedIn documentation.

File: src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationConstants.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        /// </summary>
+        public const string Email = "email";
+
+        /// <summary>
+        /// Phone number of the user.
+        /// </summary>
+        public const string PhoneNumber = "phone_number";
+
+        /// <summary>
+        /// Picture url of the user.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationConstants.cs]:
```suggestion
        /// Picture URL of the user.
```

File: src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationConstants.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        /// </summary>
+        public const string Status = "status";
+
+        /// <summary>
+        /// Verification status of the user.
+        /// </summary>
+        public const string Verified = "verified";
+
+        /// <summary>
+        /// Personal meeting url of the user.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationConstants.cs]:
```suggestion
        /// Personal meeting URL of the user.
```

File: src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationDefaults.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    /// <summary>
+    /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+    /// </summary>
+    public static readonly string TokenEndpoint = "https://zoom.us/oauth/token";
+
+    /// <summary>
+    /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
+    /// </summary>
+    public static readonly string UserInformationEndpoint = "https://zoom.us/v2/users/me";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationDefaults.cs]: According to [the documentation](https://developers.zoom.us/docs/integrations/oauth/#using-an-access-token), this should be:

```suggestion
    public static readonly string UserInformationEndpoint = "https://api.zoom.us/v2/users/me";
```

File: src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationOptions.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+using static AspNet.Security.OAuth.Zoom.ZoomAuthenticationConstants;
+
+namespace AspNet.Security.OAuth.Zoom;
+
+/// <summary>
+/// Defines a set of options used by <see cref="ZoomAuthenticationHandler"/>.
+/// </summary>
+public class ZoomAuthenticationOptions : OAuthOptions
+{
+    public ZoomAuthenticationOptions()
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationOptions.cs:17]: [According to the documentation](https://developers.zoom.us/docs/integrations/oauth/#using-a-client-app-with-pkce), Zoom supports PKCE so [`UsePkce`](https://github.com/dotnet/aspnetcore/blob/c7c76d7d3bb30e29e8386d1a3e6dae9e307fd0de/src/Security/Authentication/OAuth/src/OAuthOptions.cs#L101-L105) should be set to `true` by default.

File: src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        await Events.CreatingTicket(context);
+        return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+    }
+
+    protected override string BuildChallengeUrl([NotNull] AuthenticationProperties properties, [NotNull] string redirectUri)
+    {
+        // eBay uses the RuName for the redirect_uri
+        return base.BuildChallengeUrl(properties, redirectUri);
+    }
+
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationHandler.cs]: This appears to be unnecessary copy-paste.

```suggestion
```

File: src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationHandler.cs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zoom/ZoomAuthenticationHandler.cs]: Did you intentionally override this method to change a behaviour compared to the base class, or did you just copy-paste this from another handler?

If it is intentional, which bit is the bit that's different so I can look at the documentation to check it?

File: test/AspNet.Security.OAuth.Providers.Tests/Zoom/ZoomTests.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        });
+    }
+
+    [Theory]
+    [InlineData(ClaimTypes.NameIdentifier, "0ECPVTrOjh")]
+    [InlineData(ClaimTypes.Name, "Frodo Baggins")]
+    [InlineData(ClaimTypes.Email, "frodo@shire.middleearth")]
+    [InlineData(ClaimTypes.GivenName, "Frodo")]
+    [InlineData(ClaimTypes.Surname, "Baggins")]
+    [InlineData(ZoomAuthenticationConstants.Claims.Picture, "https://upload.wikimedia.org/wikipedia/en/4/4e/Elijah_Wood_as_Frodo_Baggins.png")]
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Zoom/ZoomTests.cs]: Please extend the test cases and the stubbed HTTP response to check each of the claims the options checks for by default.

File: test/AspNet.Security.OAuth.Providers.Tests/Zoom/ZoomTests.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        });
+    }
+
+    protected internal override void ConfigureApplication(IApplicationBuilder app)
+    {
+        app.UseRequestLocalization(new RequestLocalizationOptions
+        {
+            DefaultRequestCulture = new Microsoft.AspNetCore.Localization.RequestCulture("en-US"),
+        });
+    }
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Zoom/ZoomTests.cs]: Is this needed?

File: test/AspNet.Security.OAuth.Providers.Tests/Zoom/ZoomTests.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+public class ZoomTests(ITestOutputHelper outputHelper) : OAuthTests<ZoomAuthenticationOptions>(outputHelper)
+{
+    public override string DefaultScheme => ZoomAuthenticationDefaults.AuthenticationScheme;
+
+    protected internal override void RegisterAuthentication(AuthenticationBuilder builder)
+    {
+        builder.AddZoom(options =>
+        {
+            ConfigureDefaults(builder, options);
+        });
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Zoom/ZoomTests.cs]:
```suggestion
        builder.AddZoom(options => ConfigureDefaults(builder, options));
```

File: test/AspNet.Security.OAuth.Providers.Tests/Zoom/bundle.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+      "method": "POST",
+      "contentFormat": "json",
+      "contentJson": {
+        "access_token": "secret-access-token",
+        "expires_in": "300"
+      }
+    },
+    {
+      "comment": "https://developers.zoom.us/docs/api/rest/reference/user/methods/#tag/Users",
+      "uri": "https://zoom.us/v2/users/me",
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Zoom/bundle.json]:
```suggestion
      "uri": "https://api.zoom.us/v2/users/me",
```

File: test/AspNet.Security.OAuth.Providers.Tests/Zoom/bundle.json [hunk 13]
```diff
@@ -0,0 +1,28 @@
+{
+  "$schema": "https://raw.githubusercontent.com/justeat/httpclient-interception/master/src/HttpClientInterception/Bundles/http-request-bundle-schema.json",
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Zoom/bundle.json]:
```suggestion
  "$schema": "https://raw.githubusercontent.com/justeat/httpclient-interception/main/src/HttpClientInterception/Bundles/http-request-bundle-schema.json",
```

File: AspNet.Security.OAuth.Providers.sln [hunk 14]
```diff
@@ -219,6 +219,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "docs", "docs", "{C2CA4B38-A
 		docs\workweixin.md = docs\workweixin.md
 		docs\xumm.md = docs\xumm.md
 		docs\zendesk.md = docs\zendesk.md
+		docs\zoom.md = docs\zoom.md
```
Comment (martincostello) [inline AspNet.Security.OAuth.Providers.sln]:
```suggestion
```

**Review by martincostello — COMMENTED** [review]
This is looking good now.

Can you update the table of providers in the README to add this one too please?

Can you also confirm you've tested the provider with the real Zoom service and that the OAuth integration works as expected?

**Comment by softwareolatomiwa** [comment]
Updated the README as advised. Also, I confirm that the real Zoom service and the OAuth integration works.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for your contribution @softwareolatomiwa - this provider is now available as part of version 8.1.0.

**Pull request merged**